File: 67-broken-images/java/src/test/java/BrokenImages1.java
package src.test.java;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import net.lightbody.bmp.core.har.HarEntry;
import org.apache.commons.lang3.ObjectUtils;
import org.junit.Before;
import org.junit.After;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.Proxy;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;
import net.lightbody.bmp.BrowserMobProxy;
import net.lightbody.bmp.BrowserMobProxyServer;
import net.lightbody.bmp.client.ClientUtil;
import net.lightbody.bmp.proxy.CaptureType;
import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class BrokenImages1 {

    WebDriver driver;
    BrowserMobProxy proxy;

    @Before
    public void setUp() throws Exception {
        proxy = new BrowserMobProxyServer();
        proxy.start(0);
        Proxy seleniumProxy = ClientUtil.createSeleniumProxy(proxy);
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability(CapabilityType.PROXY, seleniumProxy);
        driver = new FirefoxDriver(capabilities);
        proxy.enableHarCaptureTypes(CaptureType.REQUEST_CONTENT,
                                    CaptureType.RESPONSE_CONTENT);
    }

    @After
    public void tearDown() throws Exception {
        proxy.stop();
        driver.quit();
    }

    @Test
    public void allImagesLoaded() throws IOException {
        proxy.newHar();
        driver.navigate().to("http://the-internet.herokuapp.com/broken_images");
        List<WebElement> images = driver.findElements(By.tagName("img"));
        List brokenImages = new ArrayList();
        List<HarEntry> harEntries = proxy.getHar().getLog().getEntries();
        for(int entry = 0; entry < harEntries.size(); entry++) {
            for(int image = 0; image < images.size(); image++) {
                if (harEntries.get(entry).getRequest().getUrl().equals(
                        images.get(image).getAttribute("src"))
                    & harEntries.get(entry).getResponse().getStatus() == 404)
                    brokenImages.add(images.get(image).getAttribute("src"));
            }
        }
        List emptyCollection = new ArrayList();
        assertThat(brokenImages, is(emptyCollection));
    }

}

File: 67-broken-images/kotlin/src/test/kotlin/BrokenImages1.kt
import net.lightbody.bmp.BrowserMobProxyServer
import net.lightbody.bmp.client.ClientUtil
import net.lightbody.bmp.proxy.CaptureType
import org.junit.jupiter.api.TestInstance
import org.openqa.selenium.By
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.firefox.FirefoxOptions
import org.openqa.selenium.firefox.FirefoxProfile
import org.openqa.selenium.remote.CapabilityType
import java.io.IOException
import kotlin.test.*

/* a single instance of of the test class is used for every method */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class BrokenImages1 {

    private val driver by lazy {
        // create an instance of Selenium connected to the proxy
        val profile = FirefoxProfile()
        // create a proxy instance and configure Selenium to use it
        val seleniumProxy = ClientUtil.createSeleniumProxy(proxy)
        val options = FirefoxOptions().setProfile(profile)
        options.setCapability(CapabilityType.PROXY, seleniumProxy)
        FirefoxDriver(options)
    }

    private val proxy by lazy {
        // start the proxy immediately
        BrowserMobProxyServer()
    }

    @BeforeTest
    @Throws(Exception::class)
    fun setUp() {

        proxy.start(0)

        // enable detailed HAR capture
        proxy.enableHarCaptureTypes(
                CaptureType.REQUEST_CONTENT,
                CaptureType.RESPONSE_CONTENT
        )
    }

    @AfterTest
    @Throws(Exception::class)
    fun tearDown() {
        proxy.stop()
        driver.quit()
    }

    @Test
    @Throws(IOException::class)
    fun `All images loaded`() {
        proxy.newHar()
        driver.navigate().to("http://the-internet.herokuapp.com/broken_images")
        val images = driver.findElements(By.tagName("img"))
        val brokenImages = ArrayList<String>()
        val harEntries = proxy.har.log.entries
        for (entry in harEntries.indices) {
            for (image in images.indices) {
                if ((harEntries[entry].request.url == images[image].getAttribute("src")) &&
                        (harEntries[entry].response.status == 404)) {
                    brokenImages.add(images[image].getAttribute("src"))
                }
            }
        }
        val emptyCollection = ArrayList<String>()
        assertEquals(
                brokenImages,
                emptyCollection
        )
    }
}